Call jit decomposition in VariableType to increase forward AD coverage

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #84151
* #84581

This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

See https://github.com/pytorch/pytorch/pull/84151#issuecomment-1238519247 for codegen output and more discussion.

soulitzer committed:
[WIP] Call jit decomposition in VariableType to increase forward AD coverage

[ghstack-poisoned]

soulitzer commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -160,6 +160,22 @@ void RegisterDecomposition(
+    }
+  }
+}
+
```
This is duplicate code from functorch

soulitzer committed:
Update on "[WIP] Call jit decomposition in VariableType to increase forward AD coverage"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Call jit decomposition in VariableType to increase forward AD coverage"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Call jit decomposition in VariableType to increase forward AD coverage"




[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/common_utils.py:
```diff
@@ -3477,3 +3477,48 @@ def custom_op(opname, symbolic_fn, opset_version):
+        graph = torch.jit.CompilationUnit(f_str).wrapped_decomp.graph
+    else:
+        graph = torch.jit.script(decomp_fn).graph
+    torch.jit._register_decomposition(decomp, graph)
```
From functorch/functorch/_src/decompositions.py

soulitzer committed:
Update on "[WIP] Call jit decomposition in VariableType to increase forward AD coverage"




[ghstack-poisoned]

soulitzer committed:
Update on "[WIP] Call jit decomposition in VariableType to increase forward AD coverage"




[ghstack-poisoned]

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name


[ghstack-poisoned]

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

samdow commented on functorch/functorch/csrc/DynamicLayer.cpp:
```diff
@@ -389,39 +389,6 @@ WithoutTop::~WithoutTop() {
-  run_jit_decomposition(op, stack);
-}
-
 static void dynamicLayerFrontFallbackOperator(
```
Could we remove the `decomp_jvp` flag in this function. It was only used for the jvp hack

soulitzer replied:
Yeah, good catch

samdow commented on functorch/test/test_ops.py:
```diff
@@ -348,13 +348,12 @@ def wrapped_fn(*args, **kwargs):
         # Composite ops that do bad things. Need to be fixed in PyTorch core.
         # RuntimeError: Cannot access data pointer of Tensor that doesn't have storage
         xfail('tensor_split'),
+        xfail('nn.functional.rrelu'),
```
nit: could we at minimum keep the comment from the previous version? It just helps us remember what the error looks like

soulitzer replied:
Whoops

samdow commented on functorch/test/test_ops.py:
```diff
@@ -1128,37 +1120,8 @@ def reference(primals, cotangents, primals_tangents, cotangents_tangents):
-            # we check that jacfwd(vjp) and jacrev(vjp) are the same. This results
-            # in slower tests.
-            FUNCTORCH_HAS_FORMULA_BUT_NOT_PYTORCH = {
-                'nn.functional.nll_loss',
-                'softmax',
-                'log_softmax',
-                'nn.functional.cross_entropy',
-                'nn.functional.layer_norm',
-                'nn.functional.batch_norm',
```
(1) 🥳 this is awesome
(2) I think I'm seeing that we're still xfailing `nll_loss`, `cross_entropy` and `batch_norm` (but not instance norm, which is just reshapes and calls batch norm 🤔 ). Just wanted to check that this matches what your seeing before I file an issue to investigate this (doesn't have to be done in this PR)

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

samdow reviewed (approved):
LGTM but we should probably have Alban and/or Richard look at the autograd codegen

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1352,13 +1385,51 @@ def emit_call(f: NativeFunction, unpacked_bindings: List[Binding]) -> str:
             guard = "at::AutoDispatchBelowADInplaceOrView guard;"
 
+        try_jit_decomposition_bool = "true" if try_jit_decomposition else "false"
+        any_has_forward_grad = (
```
Why re-do the full condition calculation instead of using the boolean flag `_any_has_forward_grad_result` ?

soulitzer replied:
We cannot use that flag because that flag is only computed if there exists a formula for that output, i.e. it looks at the formula for `result` and checks what input's tangents it uses in its computation.

albanD commented on tools/autograd/gen_variable_type.py:
```diff
@@ -1613,7 +1697,8 @@ def get_msg() -> str:
     body.append(declare_returned_variables(f))
 
-    body.append(emit_call(f, unpacked_bindings))
+    # We will need to update this part
```
To do what?

soulitzer replied:
Ohh this is out of date now. Removing

albanD commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -160,6 +160,26 @@ void RegisterDecomposition(
+  // TODO: templatize based on op and keep static trace_exec
+  auto* trace_exec = torch::jit::GetDecompositionExecutor(schema);
+  trace_exec->run((*stack));
+  if (stack->back().isTuple()) {
```
Why do we unpack tuple results here?

soulitzer replied:
Hmm not completely sure. This was from https://github.com/pytorch/functorch/pull/778, maybe @eellison knows?

eellison replied:
JIT returns a boxed tuple as a single IValue but pytorch core etc wants unboxed, multiple values on stack.

albanD reviewed (approved):
Sounds good!
Would be nice to have full example functions (for each case) from the codegen to make sure we don't re-compute stuff/duplicate code.

zou3519 commented on torch/csrc/jit/runtime/decomposition_registry.cpp:
```diff
@@ -160,6 +160,26 @@ void RegisterDecomposition(
   schema_to_decomposition[&schema] = g;
 }
 
+void run_jit_decomposition(
```
Can we delete `run_jit_decomposition` from `functorch/functorch/csrc/BatchRulesHelper.h`?

It is being used elsewhere in functorch but we should deduplicate the code

zou3519 commented:
For your codegen output:
`const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);`
- This looks wrong to me -- `findSchema` returns a value, not a reference, unless this is some c++-ism.
- findSchema is expensive, so we could cache the value by assigning it to a static variable: `static auto opt_op = ...`



```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

zou3519 reviewed (approved):
LGTM, minor nits and comments

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

soulitzer commented:
Here is an entire codegened function in VariableType:
<details>
<summary>
Click to show
</summary>

```cpp
::std::tuple<at::Tensor,at::Tensor,at::Tensor> native_batch_norm_backward(c10::DispatchKeySet ks, const at::Tensor & grad_out, const at::Tensor & input, const c10::optional<at::Tensor> & weight, const c10::optional<at::Tensor> & running_mean, const c10::optional<at::Tensor> & running_var, const c10::optional<at::Tensor> & save_mean, const c10::optional<at::Tensor> & save_invstd, bool train, double eps, ::std::array<bool,3> output_mask) {
  auto& grad_out_ = unpack(grad_out, "grad_out", 0);
  auto& input_ = unpack(input, "input", 1);
  auto _any_requires_grad = compute_requires_grad( grad_out, input, weight, save_mean, save_invstd );
  
  (void)_any_requires_grad;
  check_no_requires_grad(running_mean, "running_mean", "native_batch_norm_backward");
  check_no_requires_grad(running_var, "running_var", "native_batch_norm_backward");
  std::shared_ptr<NativeBatchNormBackwardBackward0> grad_fn;
  if (_any_requires_grad) {
    grad_fn = std::shared_ptr<NativeBatchNormBackwardBackward0>(new NativeBatchNormBackwardBackward0(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( grad_out, input, weight, save_mean, save_invstd ));
    grad_fn->grad_out_ = SavedVariable(grad_out, false);
    grad_fn->input_ = SavedVariable(input, false);
    grad_fn->weight_ = SavedVariable(weight, false);
    grad_fn->running_mean_ = SavedVariable(running_mean, false);
    grad_fn->running_var_ = SavedVariable(running_var, false);
    grad_fn->save_mean_ = SavedVariable(save_mean, false);
    grad_fn->save_invstd_ = SavedVariable(save_invstd, false);
    grad_fn->train = train;
    grad_fn->eps = eps;
  }
  at::Tensor result0;
  at::Tensor result1;
  at::Tensor result2;
  #ifndef NDEBUG
  c10::optional<Storage> grad_out__storage_saved =
    grad_out_.has_storage() ? c10::optional<Storage>(grad_out_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> grad_out__impl_saved;
  if (grad_out_.defined()) grad_out__impl_saved = grad_out_.getIntrusivePtr();
  c10::optional<Storage> input__storage_saved =
    input_.has_storage() ? c10::optional<Storage>(input_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> input__impl_saved;
  if (input_.defined()) input__impl_saved = input_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");  // Need: name and overload
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
  std::tie(result0, result1, result2) = std::move(_tmp);
  #ifndef NDEBUG
  if (grad_out__storage_saved.has_value() &&
      !at::impl::dispatch_mode_enabled() &&
      !at::impl::tensor_has_dispatch(grad_out_))
    AT_ASSERT(grad_out__storage_saved.value().is_alias_of(grad_out_.storage()));
  if (grad_out__impl_saved && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(grad_out_))
    AT_ASSERT(grad_out__impl_saved == grad_out_.getIntrusivePtr());
  if (input__storage_saved.has_value() &&
      !at::impl::dispatch_mode_enabled() &&
      !at::impl::tensor_has_dispatch(input_))
    AT_ASSERT(input__storage_saved.value().is_alias_of(input_.storage()));
  if (input__impl_saved && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(input_))
    AT_ASSERT(input__impl_saved == input_.getIntrusivePtr());
  if (result0.has_storage() && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result0)) {
    AT_ASSERT(result0.storage().use_count() == 1, "function: native_batch_norm_backward");
  }
  if (!at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result0))
    AT_ASSERT(result0.use_count() <= 1, "function: native_batch_norm_backward");
  if (result1.has_storage() && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result1)) {
    AT_ASSERT(result1.storage().use_count() == 1, "function: native_batch_norm_backward");
  }
  if (!at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result1))
    AT_ASSERT(result1.use_count() <= 1, "function: native_batch_norm_backward");
  if (result2.has_storage() && !at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result2)) {
    AT_ASSERT(result2.storage().use_count() == 1, "function: native_batch_norm_backward");
  }
  if (!at::impl::dispatch_mode_enabled() && !at::impl::tensor_has_dispatch(result2))
    AT_ASSERT(result2.use_count() <= 1, "function: native_batch_norm_backward");
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result0, result1, result2 ), grad_fn);
  }
  throw_error_for_complex_autograd(result0, "native_batch_norm_backward");
  throw_error_for_complex_autograd(result1, "native_batch_norm_backward");
  throw_error_for_complex_autograd(result2, "native_batch_norm_backward");
  return std::make_tuple(std::move(result0), std::move(result1), std::move(result2));
}
```

</details>

One thing to note is that the backward graph will be created for both the non-decomposed and decomposed levels, but this is probably OK because the graph corresponding to the decomposed operations should be destroyed as soon as we call set_history at the higher level.

An alternative to this is to actually modify codegen to no longer create the graph at this higher level when we know we are decomposition is met, but that requires more changes and increases complexity of the codegen and is probably slower because we would no longer be utilizing preexisting backward kernels.

See https://docs.google.com/document/d/1k0RLf1GrGNYLgl-Pf9s-onjjoP_wghlfFbsDxca7ImQ/edit#heading=h.bcl5pn5tsxf9 for discussion of other alternatives.

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

Example codegen output:
```cpp
  auto _tmp = ([&]() {
    if (true && (isFwGradDefined(grad_out) || isFwGradDefined(input) || isFwGradDefined(weight) || isFwGradDefined(running_mean) || isFwGradDefined(running_var) || isFwGradDefined(save_mean) || isFwGradDefined(save_invstd))) {
      c10::OperatorName full_name("aten::native_batch_norm_backward", "");
      const auto& opt_op = c10::Dispatcher::singleton().findSchema(full_name);
      TORCH_CHECK(opt_op.has_value());
      return impl::run_jit_decomposition_with_args_for_jvp<std::tuple<at::Tensor, at::Tensor, at::Tensor>, const at::Tensor &, const at::Tensor &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, const c10::optional<at::Tensor> &, bool, double, ::std::array<bool,3>>("native_batch_norm_backward", *opt_op, ks, grad_out, input, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    } else {
      at::AutoDispatchBelowADInplaceOrView guard;
      return at::redispatch::native_batch_norm_backward(ks & c10::after_autograd_keyset, grad_out_, input_, weight, running_mean, running_var, save_mean, save_invstd, train, eps, output_mask);
    }
  })();
```

[ghstack-poisoned]

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

See https://github.com/pytorch/pytorch/pull/84151#issuecomment-1238519247 for codegen output and more discussion.

[ghstack-poisoned]

soulitzer commented on tools/autograd/derivatives.yaml:
```diff
@@ -1952,7 +1952,12 @@
 - name: log_sigmoid_forward(Tensor self) -> (Tensor output, Tensor buffer)
   self: log_sigmoid_backward(grad, self, buffer)
-  output: auto_element_wise
+  # HACK: We add a `view_as` to auto_element_wise, so that we no longer have to
+  # use the jvp decomposition trick in functorch, but we also don't want
+  # test_vmapjvpall_nn_functional_logsigmoid_cuda_float32 to regress (passes on cpu!)
+  # We should either figure out what is going on with vmap or perhaps fwd AD could
+  # be more tolerant about 0-dim vs 1-dim tensors
+  output: log_sigmoid_backward(self_t.conj(), self_p, buffer).conj().view_as(self_p)
```
Forward AD was complaining about trying to set a scalar foward grad to a 1-D tensor. The reason we have this now is previously we called into the c++ composite op decomposition with `OP_DECOMPOSE`, but now we use the forward AD formula directly.

soulitzer committed:
Update on "Call jit decomposition in VariableType to increase forward AD coverage"


This PR:
- updates forward AD codegen in core to generate code that tries calling into decompositions registered to jit when
   - (1) the function is not in-place or out variant
   - AND (2) the function is differentiable (requires_derivative=True)
   - AND (3) there are no forward AD formulas registered
   - To simplify things we always generating the if/else (as long as (1) is true), but generate 'false' when either (2) or (3) are false.
 - removes the mechanism from functorch
    - (follow up) some functorch tests should be updated here so they no longer have to compute the Jacobian with vjp
  - factors out some logic to generate the any_has_forward_grad condition
     - (bc-breaking) when TensorList inputs unexpectedly have forward grad, the error will no longer contain the name

See https://github.com/pytorch/pytorch/pull/84151#issuecomment-1238519247 for codegen output and more discussion.

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

malfet commented:
@pytorchbot revert -m "Regressed test_jvpvjp_nn_functional_layer_norm_cuda_float32, see https://hud.pytorch.org/pytorch/pytorch/commit/42d99e6f196233627a28b8e9efb26a0a166fa370" -c landrace
